nvaccess/nvda
# Automatically report headers in Excel tables when using UI Automation

### Link to issue number:
None.

### Summary of the issue:
When accessing Microsoft Excel via UI Automation and moving through a table in a spreadsheet, NVDA does not automatically report row and column headers for the currently focused cell.

### Description of user facing changes
NVDA will now automatically report the row and column headers for the currently focused cell in an Excel table via UI Automation, if the row / column header is different from the last focused cell, and the spreadsheet author has enabled a header row and special first column.

### Description of development approach
* ExcelCell UIA NVDAObject: don't explicitly set columnHeaderText and rowHeaderText properties to None. This was done during prototyping and was never removed.
* UIA NVDAObject: implement the tableID property which then allows the code that reports column and row headers to correctly identify when still in the same table and therefore can stop repeting the same column and row header if it is the same as the last focused cell.

### Testing strategy:
Open [example colors Excel spreadsheet](https://github.com/nvaccess/nvda/files/9595911/example_color_table.xlsx) in Excel.
* Ensure NVDA's Access Microsoft Excel using UI Automation advanced setting is set to yes.
* Ensure that NvDA's report table headers setting in document settings panel is set to row / column
* arrow around the color table in the spreadsheet, confirming that the column header (name, red colponent, blue component) is only reported when the focus changes columns, and that the row header (purple, yellow, white etc) is only reported when the focus changes rows.

### Known issues with pull request:
None known.

### Change log entries:
New features
* Microsoft Excel via UI Automation: Automatic reporting of column and row headers in tables.
Changes
Bug fixes
For Developers

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
